### DISCUSSION TITLE
mc (midnight commander) unable to look inside archives

### DISCUSSION THREAD

**Discussion by ssbarnea** [Homebrew/discussions#3467] [Casks]
It seems that at least the current version of mc is not able to look inside archives, despite the fact that the same version from Fedora 36 does work as expected.

```
$ mc --version
GNU Midnight Commander 4.8.28
Built with GLib 2.72.0
Built with S-Lang 2.3.2 with terminfo database
Built with libssh2 1.10.0
With builtin Editor
With subshell support as default
With support for background operations
With mouse support on xterm
With internationalization support
With multiple codepages support
Virtual File Systems:
 cpiofs, tarfs, sfs, extfs, ftpfs, sftpfs, fish
Data types:
 char: 8; int: 32; long: 64; void *: 64; size_t: 64; off_t: 64;
```

Tries to press view (F3) or enter (return) on any zip file and it will not do anything about them.

**Comment by p-linnane** [comment]
Issues with the actual functioning of applications should be reported upstream.

https://www.midnight-commander.org/wiki/doc/reportDefects

**Comment by markstinson** [comment]
There was no mention if the `zip` and `unzip` utilities are installed as `mc` will open the archive in a subshell when viewing. I was able to create a zip archive and then view it with `mc` using the F3 key:
~~~
(deck@steamdeck ~)$ zip -rp dotbash.zip .bash*
  adding: .bash_history (deflated 63%)
  adding: .bash_logout (stored 0%)
  adding: .bash_profile (deflated 38%)
  adding: .bashrc (deflated 56%)
(deck@steamdeck ~)$ unzip -l dotbash.zip
Archive:  dotbash.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
     2601  2022-07-22 23:41   .bash_history
       21  2022-01-08 12:31   .bash_logout
      489  2022-05-26 16:38   .bash_profile
     2951  2022-05-26 16:38   .bashrc
---------                     -------
     6062                     4 files
(deck@steamdeck ~)$ which zip
/usr/bin/zip
(deck@steamdeck ~)$ which unzip
/usr/bin/unzip
(deck@steamdeck ~)$ which mc
/home/linuxbrew/.linuxbrew/bin/mc
(deck@steamdeck ~)$ mc
# pressed F3 on the zip archive, the Esc zero twice to exit out
(deck@steamdeck ~)$ mc --version
GNU Midnight Commander 4.8.28
Built with GLib 2.72.0
Built with S-Lang 2.3.2 with terminfo database
Built with libssh2 1.10.0
With builtin Editor
With subshell support as default
With support for background operations
With mouse support on xterm
With internationalization support
With multiple codepages support
Virtual File Systems:
 cpiofs, tarfs, sfs, extfs, ftpfs, sftpfs, fish
Data types:
 char: 8; int: 32; long: 64; void *: 64; size_t: 64; off_t: 64;
~~~

**Comment by zyv** [comment]
This is fixed in 4.8.31, caused by changes in glib.

**Reply by artemsofs** [reply]
Hello, please help how update mc in ubuntu 24.04, no way

**Reply by zyv** [reply]
Use Ubuntu avenues to ask for help: https://ubuntu.com/#community

https://bugs.launchpad.net/ubuntu/+source/mc/+bug/2062968

**Comment by artemsofs** [comment]
I installed mc 4.8.31 on ubuntu 24.04, but still can't open war files.

https://github.com/Homebrew/discussions/assets/3768326/3b288fd3-0aba-438c-9bf0-65a921c3b609

**Reply by zyv** [reply]
Well, there is no support of war files in mc.ext and there have never been:

https://github.com/MidnightCommander/mc/blob/master/misc/mc.ext.ini.in

In theory ZIP support should kick in, but it seems it's not working. What does `file` on war say, does the MIME type detection work correctly on your system?

**Reply by artemsofs** [reply]
Dear zyv, I alway use mc to work with war files. and in video you can see that mc can open war files. The problem I see is in fresh installed ubuntu 24.04.
```
mime type seems ok
root@armamax:~/releases# file --mime-type web.hospital.war 
web.hospital.war: application/java-archive
```

https://github.com/Homebrew/discussions/assets/3768326/1a3ebe26-ab17-4f16-ad62-eb269e976957

**Reply by zyv** [reply]
I don't have Ubuntu 20.04, this is a discussion area for a macOS package manager (homebrew). I asked what `file` says, not what mime-type it gives back. And I have already given you a link to the content of `mc.ext`. The relevant definitions are as follows:

```
[zip-by-type]
Type=\(Zip archive
Open=%cd %p/uzip://
View=%view{ascii} @EXTHELPERSDIR@/archive.sh view zip

[jar]
Type=\(Java\ (Jar\ file|archive)\ data\ \((zip|JAR)\)
TypeIgnoreCase=true
Open=%cd %p/uzip://
View=%view{ascii} @EXTHELPERSDIR@/archive.sh view zip
```

If this pattern is not matched because `file` output has changed, then you will not be able to enter the archives by pressing Enter.

**Reply by artemsofs** [reply]
ok Thank you

**Comment by eomer** [comment]
hello, same problem to open tar.gz files with mc 4.8.30-1build2 or 4.8.30 from universe in ubuntu 24.04.1 LTS
founded this, https://bugs.launchpad.net/ubuntu/+source/mc/+bug/2062968
seem resolved from post,  but problem stay.

**Comment by eomer** [comment]
resolved..
*** This bug is a duplicate of bug 2062968 ***
    https://bugs.launchpad.net/bugs/2062968

Workaround from #2062968:

sudo wget https://raw.githubusercontent.com/MidnightCommander/mc/master/misc/mc.ext.ini.in -O /etc/mc/mc.ext.ini

and mc reopen tar.gz to look in .. miracle! 
GREAT THANKS  to  Yury V. Zaytsev

**Reply by icebob** [reply]
It didn't work for me, but it works:

I added following to /etc/mc/mc.ext.ini file and mc started recognize .war files

```
[war-by-shell]
Shell=.war
ShellIgnoreCase=true
Open=%cd %p/uzip://
View=%view{ascii} @EXTHELPERSDIR@/archive.sh view zip
```

Thanks a lot.

**Reply by zyv** [reply]
If detection by type isn't working even after you have downloaded the latest fixed mc.ext.in from master, you must be doing something wrong, or there is another problem (related to file utility?) on your system.

The string `@EXTHELPERSDIR@` has to be replaced with the correct path to the helpers (check the file before overwriting it).

**Reply by icebob** [reply]
In this case it's strange because I didn't replace it but now it opens the war files.

**Reply by zyv** [reply]
Press F3 and observe the error message.

**Reply by artemsofs** [reply]
Dear icebob which OS you are using?

**Reply by icebob** [reply]
ubuntu 24.04

**Reply by icebob** [reply]
> Press F3 and observe the error message.

Yeah, I saw. On F3 it shows error, but with ENTER it opens the war.

**Reply by artemsofs** [reply]
[mc.ext.ini.txt](https://github.com/user-attachments/files/17677945/mc.ext.ini.txt)
Just replace your mc.ext.ini file information with attached file information, and all will be work ok
/etc/mc/mc.ext.ini

**Reply by icebob** [reply]
Thx @artemsofs, entering war works with your mc.ext.ini as well!
However F3 drops the same error message.

**Reply by zyv** [reply]
I've told you guys to replace `@EXTHELPERSDIR@` with `/usr/lib/mc/ext.d`, but you are not reading :(

**Reply by rPraml** [reply]
Thanks @zyv `sudo sed -e 's/@EXTHELPERSDIR@/\/usr\/lib\/mc\/ext.d/g' /etc/mc/mc.ext.ini -i` does the rest
